devel/management/commands/reposec.py

```diff
@@ -0,0 +1,237 @@
+    format='%(asctime)s -> %(levelname)s: %(message)s',
+    datefmt='%Y-%m-%d %H:%M:%S',
+    stream=sys.stderr)
+TRACE = 5
+logging.addLevelName(TRACE, 'TRACE')
+logger = logging.getLogger()
+
+class Command(BaseCommand):
+    help = ""
+    missing_args_message = 'missing arch and file.'
+
+    def add_arguments(self, parser):
+        parser.add_argument('args', nargs='*', help='<arch> <filename>')
+        parser.add_argument('--processes',
+                            action='store_true',
+                            dest='processes',
+                            default=cpu_count(),
+                            help=f'number of parallel processes (default: {cpu_count()})')
+
+
+    def handle(self, arch=None, directory=None, processes=cpu_count(), **options):
+        if not arch:
+            raise CommandError('Architecture is required.')
+        if not directory:
+            raise CommandError('Repo location is required.')
+        directory = os.path.normpath(directory)
+        if not os.path.exists(directory):
+            raise CommandError('Specified repository location does not exists.')
+
+        v = int(options.get('verbosity', 0))
+        if v == 0:
+            logger.level = logging.ERROR
+        elif v == 1:
+            logger.level = logging.INFO
+        elif v >= 2:
+            logger.level = logging.DEBUG
+
+        return read_repo(arch, directory, processes, options)
+
+
+def read_file(arch, repo, filename):
+    pkgsec = None
+    basename = os.path.basename(filename)
+    pkgname = basename.rsplit('-', 3)[0]
+
+    with file_reader(filename) as pkg:
+        for entry in pkg:
+            if not entry.isfile:
+                continue
+
+            # Retrieve pkgname
+            if entry.name == '.PKGINFO':
+                continue
+
+            if not entry.name.startswith('usr/bin/'):
+                continue
+
+            fp = io.BytesIO(b''.join(entry.get_blocks()))
+            elf = Elf(fp)
+
+            if not elf.is_elf():
+                continue
+
+            pkg = Package.objects.get(pkgname=pkgname, repo=repo)
+            pkgsec = PackageSecurity(pkg=pkg, is_pie=elf.pie, is_relro=elf.relro, is_canary=elf.canary)
+
+    return pkgsec
+
+
+
+def read_repo(arch, source_dir, processes, options):
+    tasks = []
+
+    directory = os.path.join(source_dir, 'os', arch)
+    for filename in glob(os.path.join(directory, f'*{PKG_EXT}')):
+        tasks.append((filename))
+
+    arch = Arch.objects.get(name=arch)
+
+    reponame = os.path.basename(source_dir).title()
+    repo = Repo.objects.get(name=reponame)
+
+
+    with Pool(processes=processes) as pool:
+        results = pool.map(partial(read_file, arch, repo), tasks)
+
+    results = [r for r in results if r]
+    results = [r for r in results if not r.is_pie or not r.is_relro or not r.is_canary]
+
+    with transaction.atomic():
+        PackageSecurity.objects.all().delete()
+        PackageSecurity.objects.bulk_create(results)
```

Comment:
Not sure if it makes a big difference, but in `reporead` we:
* Create ones that exist in the repo but not in the DB.
* Delete ones in the DB but not in the repo.
* Update ones that are in both.

I don't know how big the difference in performance is, but given how many packages we have, it wouldn't surprise me if it gets somewhat substantial.

If we want to do it in a similar way, I think something like this could work:

* Before reading the repo, get a list of all package names and versions for that repo from the DB, probably in a `{pkgname: pkgver}` dict.

* Iterate over each package in the pacman repo file:
  * `if pkgname not in pkgdict` -> add the package file path to the `to_add` list.
  * `if pkgname in pkgdict`
    * and the version is the same, pop the package name from the package dict.
    * and the version is different, leave the package in the dict so it gets deleted and also add the package file path to the `to_add` list. 

* Delete everything left in the package dict from the repo.
* Parse all package files in the `to_add` list and add them to the repo.

Reply:
(But if this turns out to be fairly quick on our repos as it is, feel free to ignore this.)

Reply:
This sounds smart! Let me try to implement something like it.

Reply:
I have only tried [core] and I think we should do this efficient. Also currently it removes all PackageSecurity's while it could just remove for Arch.

Reply:
So I'm not so sure how this would work. We don't have a list of parsed packages since this implementation only adds a PackageSecurity object for once a file has a "violation". This kinda sucks for [community] which is huge. So maybe this needs some re-designing or re-use the PackageFile object.